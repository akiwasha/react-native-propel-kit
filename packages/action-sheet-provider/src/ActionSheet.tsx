import ModalDialog, {ModalDialogHandle, ModalDialogProps} from '@mgcrea/react-native-modal-dialog';
import React, {forwardRef, ForwardRefRenderFunction, useMemo} from 'react';
import {
  ActionSheetIOSOptions,
  Dimensions,
  Platform,
  ScrollView,
  ScrollViewProps,
  StyleProp,
  StyleSheet,
  TextStyle,
  ViewStyle
} from 'react-native';
import ActionSheetOption from './components/ActionSheetOption';
import isUndefined from './utils/isUndefined';

const {height: WINDOW_HEIGHT} = Dimensions.get('window');
const IOS_OPACITY = 0.85; // @NOTE from native ActionSheet (though it uses blur)

export type ActionSheetProps = Pick<ScrollViewProps, 'scrollEnabled'> &
  Omit<ModalDialogProps, 'onConfirm' | 'confirmTitle' | 'confirmStyle'> &
  ActionSheetIOSOptions & {
    defaultStyles?: typeof defaultStyles;
    optionStyle?: StyleProp<TextStyle>;
    lastOptionExtraStyle?: StyleProp<TextStyle>;
    destructiveButtonColor?: string;
    onButtonPress: (buttonIndex: number) => void;
    onCancel: () => void;
  };

export const defaultProps = {
  backgroundColor: 'white',
  cancelTitle: 'Cancel',
  destructiveButtonColor: Platform.select({
    ios: '#ff3b30', // iOS.systemRed (@see https://developer.apple.com/design/human-interface-guidelines/ios/visual-design/color/)
    android: '#f44336' // android.red500 (@see https://material.io/design/color/the-color-system.html)
  })!
};

const ActionSheetRenderFunction: ForwardRefRenderFunction<ModalDialogHandle, ActionSheetProps> = (
  {
    options,
    backgroundColor = defaultProps.backgroundColor,
    optionStyle: propOptionStyle,
    lastOptionExtraStyle: propLastOptionExtraStyle,
    cancelButtonIndex,
    destructiveButtonIndex,
    destructiveButtonColor = defaultProps.destructiveButtonColor,
    onButtonPress,
    scrollEnabled: propScrollEnabled,
    defaultStyles: propDefaultStyles = defaultStyles,
    ...otherModalDialogProps
  },
  ref
) => {
  const cancelTitle = !isUndefined(cancelButtonIndex) && options[cancelButtonIndex] ? options[cancelButtonIndex] : '';
  const scrollEnabled = !isUndefined(cancelButtonIndex) ? options.length > 6 : propScrollEnabled;

  const optionStyle = useMemo<ViewStyle>(
    () => StyleSheet.flatten([propDefaultStyles.option, {backgroundColor}, propOptionStyle]),
    [propDefaultStyles, backgroundColor, propOptionStyle]
  );
  const lastOptionExtraStyle = useMemo<ViewStyle>(
    () => StyleSheet.flatten([propDefaultStyles.lastOptionExtra, propLastOptionExtraStyle]),
    [propDefaultStyles, propLastOptionExtraStyle]
  );

  return (
    <ModalDialog
      ref={ref}
      defaultStyles={propDefaultStyles}
      backgroundColor={backgroundColor}
      cancelTitle={cancelTitle}
      bodyStyle={{backgroundColor: 'transparent'}}
      {...otherModalDialogProps}>
      <ScrollView scrollEnabled={scrollEnabled}>
        {options.map((buttonTitle, buttonIndex) => {
          // Skip cancel button
          if (!isUndefined(cancelButtonIndex) && buttonIndex === cancelButtonIndex) {
            return null;
          }

          const color: string =
            destructiveButtonIndex && buttonIndex === destructiveButtonIndex ? destructiveButtonColor : '';
          return (
            <ActionSheetOption
              style={[optionStyle as TextStyle].concat(
                color ? [{color}] : [],
                buttonIndex === options.length - 1 ? [lastOptionExtraStyle] : []
              )}
              onPress={() => onButtonPress(buttonIndex)}
              title={buttonTitle}
              key={`${buttonIndex}`} // eslint-disable-line react/no-array-index-key
            />
          );
        })}
      </ScrollView>
    </ModalDialog>
  );
};

export const ActionSheet = forwardRef(ActionSheetRenderFunction);

export const defaultStyles = {
  modal: Platform.select<ViewStyle>({
    ios: {flex: 1, justifyContent: 'flex-end'},
    android: {flex: 1, justifyContent: 'center'}
  })!,
  container: Platform.select<ViewStyle>({
    ios: {
      flexDirection: 'column',
      alignItems: 'stretch',
      marginHorizontal: 8,
      marginBottom: 8,
      maxHeight: WINDOW_HEIGHT - 8
    },
    android: {
      flexDirection: 'column',
      alignItems: 'stretch',
      // marginHorizontal: 82, // timepicker/spinner
      margin: 24,
      maxHeight: WINDOW_HEIGHT - 24 * 2,
      elevation: 4
    }
  })!,
  header: Platform.select<ViewStyle>({
    ios: {
      alignItems: 'center',
      padding: 16,
      borderTopLeftRadius: 12,
      borderTopRightRadius: 12,
      opacity: IOS_OPACITY
    },
    android: {alignItems: 'flex-start', paddingVertical: 24, paddingHorizontal: 32}
  })!,
  body: Platform.select<ViewStyle>({
    ios: {
      flexShrink: 1,
      flexDirection: 'column',
      backgroundColor: 'transparent',
      opacity: IOS_OPACITY,
      borderBottomStartRadius: 12,
      borderBottomEndRadius: 12
    },
    android: {
      flexShrink: 1,
      flexDirection: 'column',
      backgroundColor: 'transparent'
    }
  })!,
  footer: Platform.select<TextStyle>({
    ios: {}, // @NOTE we don't have a footer on iOS
    android: {
      flexDirection: 'row',
      justifyContent: 'flex-end',
      borderBottomLeftRadius: 2,
      borderBottomRightRadius: 2,
      padding: 8
    }
  })!,
  title: Platform.select<TextStyle>({
    ios: {paddingBottom: 12, fontSize: 14, fontWeight: '500', textAlign: 'center', color: '#888'},
    android: {
      paddingBottom: 12,
      fontSize: 20,
      fontWeight: '500',
      textAlign: 'left',
      color: '#333'
    }
  })!,
  message: Platform.select<TextStyle>({
    ios: {paddingBottom: 12, fontSize: 13, fontWeight: '400', textAlign: 'center', color: '#888'},
    android: {paddingBottom: 12, fontSize: 18, fontWeight: '400', color: '#666'}
  })!,
  cancel: Platform.select<TextStyle>({
    ios: {marginTop: 8, borderRadius: 12, fontWeight: '600'},
    android: {}
  })!,
  confirm: Platform.select<TextStyle>({
    ios: {}, // @NOTE we don't have a confirm on iOS
    android: {}
  })!,
  option: Platform.select<TextStyle>({
    ios: {
      marginTop: StyleSheet.hairlineWidth
    },
    android: {
      marginTop: StyleSheet.hairlineWidth
    }
  })!,
  lastOptionExtra: Platform.select<TextStyle>({
    ios: {
      borderBottomLeftRadius: 12,
      borderBottomRightRadius: 12
    },
    android: {
      marginBottom: StyleSheet.hairlineWidth
    }
  })!
};
