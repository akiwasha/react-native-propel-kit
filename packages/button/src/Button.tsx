// @see https://github.com/facebook/react-native/blob/master/Libraries/Components/Button.js
// @see https://github.com/react-native-training/react-native-elements/blob/next/src/buttons/Button.js

import React, {ElementType, FunctionComponent, useCallback, useMemo} from 'react';
import {
  ButtonProps as BaseButtonProps,
  GestureResponderEvent,
  Platform,
  StyleProp,
  StyleSheet,
  Text,
  TextStyle,
  TouchableNativeFeedback,
  TouchableNativeFeedbackProps,
  TouchableOpacity,
  TouchableOpacityProps,
  View,
  ViewStyle
} from 'react-native';
import {pickTextStyles} from './utils';

export const defaultStyles = {
  view: Platform.select<ViewStyle>({
    ios: {justifyContent: 'center', flexDirection: 'row', height: 50},
    android: {justifyContent: 'center', flexDirection: 'row', height: 50}
  }),
  disabledView: Platform.select<ViewStyle>({
    ios: {},
    android: {}
  }),
  text: Platform.select<TextStyle>({
    ios: {textAlign: 'center', margin: 8, color: '#007aff', fontSize: 18},
    android: {paddingHorizontal: 3, paddingBottom: 2, color: '#212121'}
  }),
  disabledText: Platform.select<TextStyle>({
    ios: {color: '#cdcdcd'},
    android: {color: '#cdcdcd'}
  })
  // loading: Platform.select<ViewStyle>({
  //   ios: {
  //     position: 'absolute',
  //     backgroundColor: 'black',
  //     opacity: 0.5,
  //     height: 40,
  //     width: 40,
  //     marginRight: 8,
  //     alignSelf: 'center',
  //     borderRadius: 12
  //   },
  //   android: {marginRight: 4}
  // })
};

export type ButtonProps = BaseButtonProps & {
  // loading?: boolean;
  // loadingStyle?: StyleProp<ViewStyle>;
  style?: StyleProp<TextStyle>;
  viewStyle?: StyleProp<ViewStyle>;
  disabledViewStyle?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
  disabledTextStyle?: StyleProp<TextStyle>;
  TouchableComponent?: ElementType<TouchableNativeFeedbackProps | TouchableOpacityProps>;
  activeOpacity?: TouchableOpacityProps['activeOpacity'];
  defaultStyles?: typeof defaultStyles;
};

export const defaultProps = {
  TouchableComponent: Platform.select<ElementType<TouchableNativeFeedbackProps | TouchableOpacityProps>>({
    android: TouchableNativeFeedback,
    default: TouchableOpacity
  }),
  defaultStyles: typeof defaultStyles
};

export const Button: FunctionComponent<ButtonProps> = ({
  title,
  disabled,
  // loading: propLoading,
  onPress,
  color,
  viewStyle: propViewStyle,
  // loadingStyle: propLoadingStyle,
  disabledViewStyle,
  textStyle: propTextStyle,
  disabledTextStyle,
  defaultStyles: propDefaultStyles = defaultStyles,
  style: propStyle,
  TouchableComponent = defaultProps.TouchableComponent,
  ...otherTouchableProps
}) => {
  // @NOTE we use flatten to properly split text/view related styles
  const flattenStyle = useMemo<TextStyle>(() => (propStyle ? StyleSheet.flatten(propStyle) : {}), [propStyle]);

  const handlePress = useCallback(
    (ev: GestureResponderEvent) => {
      if (disabled) {
        return;
      }
      if (onPress) {
        onPress(ev);
      }
    },
    [disabled, onPress]
  );

  const viewStyle = useMemo<ViewStyle>(
    () =>
      StyleSheet.flatten([
        propDefaultStyles.view,
        propViewStyle,
        flattenStyle,
        disabled ? [propDefaultStyles.disabledView, disabledViewStyle] : null
      ]),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [propViewStyle, flattenStyle, disabled, disabledViewStyle]
  );

  const textStyle = useMemo<TextStyle>(
    () =>
      StyleSheet.flatten([
        propDefaultStyles.text,
        propTextStyle,
        pickTextStyles(flattenStyle),
        color ? {color} : null,
        disabled ? [propDefaultStyles.disabledText, disabledTextStyle] : null
      ]),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [propTextStyle, flattenStyle, disabled, color, disabledTextStyle]
  );

  return (
    <TouchableComponent
      style={{flex: viewStyle.flex ? 1 : 0, flexDirection: 'row'}}
      disabled={disabled}
      onPress={handlePress}
      {...otherTouchableProps}
    >
      <View style={viewStyle}>
        {/* {propLoading ? (
          <ActivityIndicator
            style={[propDefaultStyles.loading, propLoadingStyle]}
            color={textStyle && textStyle.color ? textStyle.color : propDefaultStyles.text.color}
          />
        ) : null} */}
        <Text style={textStyle}>{title}</Text>
      </View>
    </TouchableComponent>
  );
};
