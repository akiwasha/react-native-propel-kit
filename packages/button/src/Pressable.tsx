// @see https://github.com/facebook/react-native/blob/master/Libraries/Components/Button.js
// @see https://github.com/react-native-training/react-native-elements/blob/next/src/buttons/Button.js

import React, {ElementType, FunctionComponent, ReactNode, useCallback, useMemo} from 'react';
import {
  ButtonProps,
  GestureResponderEvent,
  Platform,
  StyleProp,
  StyleSheet,
  Text,
  TextProps,
  TextStyle,
  TouchableNativeFeedback,
  TouchableNativeFeedbackProps,
  TouchableOpacity,
  TouchableOpacityProps,
  ViewStyle
} from 'react-native';
import {pickTextStyles} from './utils';

export type PressableProps = ButtonProps &
  Pick<TextProps, 'ellipsizeMode' | 'numberOfLines'> & {
    children?: ReactNode;
    style?: StyleProp<TextStyle>;
    viewStyle?: StyleProp<ViewStyle>;
    disabledViewStyle?: StyleProp<ViewStyle>;
    textStyle?: StyleProp<TextStyle>;
    disabledTextStyle?: StyleProp<TextStyle>;
    TouchableComponent?: ElementType<TouchableNativeFeedbackProps | TouchableOpacityProps>;
    activeOpacity?: TouchableOpacityProps['activeOpacity'];
    defaultStyles?: typeof defaultStyles;
  };

export const defaultProps = {
  TouchableComponent: Platform.select<ElementType<TouchableNativeFeedbackProps | TouchableOpacityProps>>({
    android: TouchableNativeFeedback,
    default: TouchableOpacity
  })
};

export const Pressable: FunctionComponent<PressableProps> = ({
  children,
  title,
  disabled,
  onPress,
  color,
  viewStyle: propViewStyle,
  disabledViewStyle,
  textStyle: propTextStyle,
  ellipsizeMode,
  numberOfLines,
  disabledTextStyle,
  defaultStyles: propDefaultStyles = defaultStyles,
  style: propStyle,
  TouchableComponent = defaultProps.TouchableComponent,
  ...otherTouchableProps
}) => {
  // @NOTE we use flatten to properly split text/view related styles
  const flattenStyle = useMemo<TextStyle>(() => (propStyle ? StyleSheet.flatten(propStyle) : {}), [propStyle]);

  const handlePress = useCallback(
    (ev: GestureResponderEvent) => {
      if (disabled) {
        return;
      }
      if (onPress) {
        onPress(ev);
      }
    },
    [disabled, onPress]
  );

  const viewStyle = useMemo<ViewStyle>(
    () =>
      StyleSheet.flatten([propDefaultStyles.view, propViewStyle, flattenStyle, disabled ? disabledViewStyle : null]),
    [propDefaultStyles.view, propViewStyle, flattenStyle, disabled, disabledViewStyle]
  );

  const textStyle = useMemo<TextStyle>(
    () =>
      StyleSheet.flatten([
        propDefaultStyles.text,
        propTextStyle,
        pickTextStyles(flattenStyle),
        color ? {color} : null,
        disabled ? disabledTextStyle : null
      ]),
    [propDefaultStyles.text, propTextStyle, flattenStyle, disabled, color, disabledTextStyle]
  );

  return (
    <TouchableComponent style={viewStyle} disabled={disabled} onPress={handlePress} {...otherTouchableProps}>
      {children}
      <Text ellipsizeMode={ellipsizeMode} numberOfLines={numberOfLines} style={textStyle}>
        {title}
      </Text>
    </TouchableComponent>
  );
};

export const defaultStyles = {
  view: Platform.select<ViewStyle>({
    ios: {justifyContent: 'center', flexDirection: 'row', height: 50},
    android: {justifyContent: 'center', flexDirection: 'row', height: 50}
  }),
  disabledView: Platform.select<ViewStyle>({
    ios: {},
    android: {}
  }),
  text: Platform.select<TextStyle>({
    ios: {alignSelf: 'center', fontSize: 18, color: '#007aff'},
    android: {paddingHorizontal: 3, paddingBottom: 2, color: '#212121'}
  }),
  disabledText: Platform.select<TextStyle>({
    ios: {color: '#cdcdcd'},
    android: {color: '#cdcdcd'}
  })
};
