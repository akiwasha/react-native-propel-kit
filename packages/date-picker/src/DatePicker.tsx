import {InputButton, InputButtonProps} from '@mgcrea/react-native-button';
import ModalDialog, {ModalDialogHandle, ModalDialogProps} from '@mgcrea/react-native-modal-dialog';
import DateTimePicker, {AndroidNativeProps, IOSNativeProps} from '@react-native-community/datetimepicker';
import React, {
  ElementType,
  forwardRef,
  useCallback,
  useEffect,
  useImperativeHandle,
  useMemo,
  useRef,
  useState
} from 'react';
import {Platform, View} from 'react-native';
import {
  asLocaleDate,
  asStartOfDate,
  asUTCDate,
  DatePickerAndroid,
  defaultLabelExtractor,
  getNavigatorLocale,
  isUndefined,
  LabelExtractorOptions,
  openAndroidDatePicker,
  trimDate
} from './utils';

// const CURRENT_YEAR = new Date().getFullYear();
// const FIRST_DAY_OF_YEAR = new Date(Date.UTC(CURRENT_YEAR, 0, 1));
const START_OF_TODAY = asStartOfDate(new Date(), 'day');

export type DatePickerHandle = {
  focus: InputButtonProps['onFocus'];
};

export type DatePickerProps = Pick<ModalDialogProps, 'title' | 'confirmTitle' | 'cancelTitle'> &
  Pick<IOSNativeProps, 'mode' | 'display' | 'locale'> & {
    androidDisplay?: AndroidNativeProps['display'];
    initialValue?: Date;
    InputButtonComponent?: ElementType<InputButtonProps>;
    labelExtractor?: (value: Date, options: LabelExtractorOptions) => string;
    onChange?: (value: Date) => void;
    placeholder?: string;
    onSubmitEditing?: (value: Date) => void; // Callback that is called when the text input's submit button is pressed.
    value?: Date;
    utc?: boolean;
    trim?: boolean;
    [s: string]: unknown; // otherInputButtonProps
  };

export const defaultProps: Required<
  Pick<
    DatePickerProps,
    | 'androidDisplay'
    | 'initialValue'
    | 'InputButtonComponent'
    | 'labelExtractor'
    | 'locale'
    | 'mode'
    | 'display'
    | 'utc'
    | 'trim'
  >
> = {
  androidDisplay: 'default',
  initialValue: START_OF_TODAY,
  InputButtonComponent: InputButton,
  labelExtractor: defaultLabelExtractor,
  locale: getNavigatorLocale(),
  mode: 'date',
  display: 'inline',
  utc: false,
  trim: false
};

export const DatePicker = forwardRef<DatePickerHandle, DatePickerProps>(
  (
    {
      // ModalDialog props
      cancelTitle,
      confirmTitle,
      title,
      // InputButton props
      placeholder,
      // DatePicker props
      mode = defaultProps.mode,
      InputButtonComponent = defaultProps.InputButtonComponent,
      androidDisplay = defaultProps.androidDisplay,
      initialValue = defaultProps.initialValue,
      labelExtractor = defaultProps.labelExtractor,
      locale = defaultProps.locale,
      onChange: propOnChange,
      value: propValue,
      utc = defaultProps.utc,
      trim = defaultProps.trim,
      display = defaultProps.display,
      onSubmitEditing,
      ...otherInputButtonProps
    },
    ref
  ) => {
    const value = propValue || initialValue;
    const [modalValue, setModalValue] = useState<Date>(value);

    // Track parent propValue controlled updates
    useEffect(() => {
      const nextValue = utc ? asLocaleDate(value) : value;
      setModalValue(nextValue);
    }, [value, utc]);

    // Support both controlled/uncontrolled usage
    const [localValue, setLocalValue] = useState<Date>(initialValue);
    const inputValue = !isUndefined(propValue) ? propValue : localValue;

    // Lazily compute displayed label
    const labelValue = useMemo<string>(() => {
      if (isUndefined(inputValue)) {
        return '';
      }
      if (labelExtractor) {
        return labelExtractor(inputValue, {mode, locale, utc});
      }
      return `${inputValue}`;
    }, [labelExtractor, inputValue, mode, locale, utc]);

    const computeDate = (date: Date = initialValue) => {
      let nextValue = date;
      // console.warn({nextValue});
      if (trim) {
        nextValue = trimDate(nextValue, mode);
        // console.warn({trim: nextValue});
      }
      if (utc) {
        nextValue = asUTCDate(nextValue);
        // console.warn({utc: nextValue});
      }
      return nextValue;
    };

    // Propagate changed value
    const onConfirm = useCallback(
      (value?: Date) => {
        // @NOTE android uses direct calls while ios relies on a modalValue
        const nextValue = computeDate(Platform.select({ios: modalValue, android: value}));
        if (propOnChange) {
          propOnChange(nextValue);
        }
        if (onSubmitEditing) {
          // @NOTE prevent swallowing siblings focus events
          setTimeout(() => {
            onSubmitEditing(nextValue);
          });
        }
        // Support uncontrolled usage
        if (isUndefined(propValue)) {
          setLocalValue(nextValue);
        }
      },
      [mode, utc, propValue, modalValue, propOnChange, onSubmitEditing]
    );

    // Reset modalValue to the proper value
    const onCancel = useCallback(() => {
      setModalValue(inputValue || initialValue);
    }, [inputValue]);

    // Open the modal when user touches input
    const modalDialogRef = useRef<ModalDialogHandle>(null);
    const focus = Platform.select<InputButtonProps['onFocus']>({
      android: useCallback(async () => {
        const {action, value: nextValue} = await openAndroidDatePicker(mode, {
          value: inputValue,
          display: androidDisplay,
          utc
        });
        if (action !== DatePickerAndroid.dismissedAction) {
          onConfirm(computeDate(nextValue));
        }
      }, [inputValue, mode, onConfirm]),
      default: useCallback(() => {
        if (modalDialogRef.current) {
          modalDialogRef.current.show();
        }
      }, [])
    });
    // Expose API via an imperative handle
    useImperativeHandle<DatePickerHandle, DatePickerHandle>(ref, () => ({focus}), [focus]);

    return (
      <>
        <InputButtonComponent onFocus={focus} placeholder={placeholder} value={labelValue} {...otherInputButtonProps} />
        {Platform.OS === 'ios' ? (
          <ModalDialog
            ref={modalDialogRef}
            title={title || placeholder}
            onConfirm={onConfirm}
            onCancel={onCancel}
            confirmTitle={confirmTitle}
            cancelTitle={cancelTitle}
          >
            <View style={{height: mode === 'datetime' ? 417.5 : 368, width: '100%'}}>
              <DateTimePicker
                value={modalValue}
                mode={mode}
                style={{flex: 1}}
                display={display}
                onChange={(_event, date) => {
                  if (date) {
                    setModalValue(date);
                  }
                }}
              />
            </View>
          </ModalDialog>
        ) : null}
      </>
    );
  }
);

// 368
//417.5
