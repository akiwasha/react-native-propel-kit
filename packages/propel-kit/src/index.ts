export * from '@mgcrea/react-native-action-sheet-provider';
export * from '@mgcrea/react-native-backdrop-provider';
export * from '@mgcrea/react-native-button';
export * from '@mgcrea/react-native-date-picker';
export * from '@mgcrea/react-native-modal-dialog';
export * from '@mgcrea/react-native-select';
