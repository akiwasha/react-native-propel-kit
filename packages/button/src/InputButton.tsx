import React, {FunctionComponent} from 'react';
import {Platform, StyleProp, TextInputProps, TextStyle, ViewStyle} from 'react-native';
import {Pressable, PressableProps} from './Pressable';

export type InputButtonProps = Omit<TextInputProps, 'onPress'> & {
  onFocus: PressableProps['onPress'];
  viewStyle?: StyleProp<ViewStyle>;
  disabledViewStyle?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
  disabledTextStyle?: StyleProp<TextStyle>;
};

export const defaultProps = {
  placeholderTextColor: Platform.select({ios: '#c7c7cc', android: '#9e9e9e'})
};

export const InputButton: FunctionComponent<InputButtonProps> = ({
  editable = true,
  onFocus,
  placeholder,
  viewStyle,
  disabledViewStyle,
  textStyle,
  disabledTextStyle,
  placeholderTextColor = defaultProps.placeholderTextColor,
  value,
  ...otherPressableProps
}) => {
  return (
    <Pressable
      onPress={onFocus}
      title={value || placeholder || ''}
      disabled={!editable}
      viewStyle={[defaultStyles.view, viewStyle]}
      disabledViewStyle={[defaultStyles.disabledView, disabledViewStyle]}
      textStyle={[defaultStyles.text, textStyle]}
      disabledTextStyle={[defaultStyles.disabledText, disabledTextStyle]}
      color={!value ? placeholderTextColor : undefined}
      numberOfLines={1}
      {...otherPressableProps}
    />
  );
};

export const defaultStyles: {[s: string]: ViewStyle | TextStyle} = {
  view: Platform.select<ViewStyle>({
    ios: {justifyContent: 'flex-start', alignItems: 'center'},
    android: {justifyContent: 'flex-start', height: 50}
  })!,
  disabledView: Platform.select<ViewStyle>({
    ios: {},
    android: {}
  })!,
  text: Platform.select<TextStyle>({
    ios: {fontSize: 14, color: 'black'},
    android: {alignSelf: 'flex-start', paddingHorizontal: 3, paddingBottom: 2, color: '#212121'}
  })!,
  disabledText: Platform.select<TextStyle>({
    ios: {color: '#cdcdcd'},
    android: {color: '#cdcdcd'}
  })!
};
