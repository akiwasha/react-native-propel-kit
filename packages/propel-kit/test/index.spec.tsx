import React from 'react';

import * as exportedInterface from '../src';

describe('propel-kit', () => {
  it('export a proper interface', () => {
    expect(exportedInterface).toMatchSnapshot();
  });
});
