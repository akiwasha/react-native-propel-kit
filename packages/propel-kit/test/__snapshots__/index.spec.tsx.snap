// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`propel-kit export a proper interface 1`] = `
Object {
  "ActionSheet": Object {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "ActionSheetContext": Object {
    "$$typeof": Symbol(react.context),
    "Consumer": Object {
      "$$typeof": Symbol(react.context),
      "_calculateChangedBits": null,
      "_context": [Circular],
    },
    "Provider": Object {
      "$$typeof": Symbol(react.provider),
      "_context": [Circular],
    },
    "_calculateChangedBits": null,
    "_currentRenderer": null,
    "_currentRenderer2": null,
    "_currentValue": null,
    "_currentValue2": null,
    "_threadCount": 0,
  },
  "ActionSheetProvider": [Function],
  "ActionSheetProviderDefaultProps": Object {
    "native": true,
  },
  "BackdropContext": Object {
    "$$typeof": Symbol(react.context),
    "Consumer": Object {
      "$$typeof": Symbol(react.context),
      "_calculateChangedBits": null,
      "_context": [Circular],
    },
    "Provider": Object {
      "$$typeof": Symbol(react.provider),
      "_context": [Circular],
    },
    "_calculateChangedBits": null,
    "_currentRenderer": null,
    "_currentRenderer2": null,
    "_currentValue": null,
    "_currentValue2": null,
    "_threadCount": 0,
  },
  "BackdropProvider": [Function],
  "Button": [Function],
  "DatePicker": Object {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "DateTimePicker": Object {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "InputButton": [Function],
  "ModalDialog": Object {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "ModalDialogButton": [Function],
  "MonthPicker": Object {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "Pressable": [Function],
  "Select": Object {
    "$$typeof": Symbol(react.forward_ref),
    "Item": [Function],
    "render": [Function],
  },
  "SelectItem": [Function],
  "TimePicker": Object {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "YearPicker": Object {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "actionSheetDefaultProps": Object {
    "backgroundColor": "white",
    "cancelTitle": "Cancel",
    "destructiveButtonColor": "#ff3b30",
  },
  "actionSheetDefaultStyles": Object {
    "body": Object {
      "backgroundColor": "transparent",
      "borderBottomEndRadius": 12,
      "borderBottomStartRadius": 12,
      "flexDirection": "column",
      "flexShrink": 1,
      "opacity": 0.85,
    },
    "cancel": Object {
      "borderRadius": 12,
      "fontWeight": "600",
      "marginTop": 8,
    },
    "confirm": Object {},
    "container": Object {
      "alignItems": "stretch",
      "flexDirection": "column",
      "marginBottom": 8,
      "marginHorizontal": 8,
      "maxHeight": 1326,
    },
    "footer": Object {},
    "header": Object {
      "alignItems": "center",
      "borderTopLeftRadius": 12,
      "borderTopRightRadius": 12,
      "opacity": 0.85,
      "padding": 16,
    },
    "lastOptionExtra": Object {
      "borderBottomLeftRadius": 12,
      "borderBottomRightRadius": 12,
    },
    "message": Object {
      "color": "#888",
      "fontSize": 13,
      "fontWeight": "400",
      "paddingBottom": 12,
      "textAlign": "center",
    },
    "modal": Object {
      "flex": 1,
      "justifyContent": "flex-end",
    },
    "option": Object {
      "marginTop": 0.5,
    },
    "title": Object {
      "color": "#888",
      "fontSize": 14,
      "fontWeight": "500",
      "paddingBottom": 12,
      "textAlign": "center",
    },
  },
  "backdropProviderDefaultProps": Object {
    "backgroundColor": "black",
    "duration": 300,
    "easing": [Function],
    "opacity": 0.25,
    "useNativeDriver": true,
    "zIndex": 99,
  },
  "buttonDefaultProps": Object {
    "TouchableComponent": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "defaultStyles": "object",
  },
  "buttonDefaultStyles": Object {
    "disabledText": Object {
      "color": "#cdcdcd",
    },
    "disabledView": Object {},
    "text": Object {
      "color": "#007aff",
      "fontSize": 18,
      "margin": 8,
      "textAlign": "center",
    },
    "view": Object {
      "flexDirection": "row",
      "height": 50,
      "justifyContent": "center",
    },
  },
  "datePickerDefaultProps": Object {
    "InputButtonComponent": [Function],
    "androidDisplay": "default",
    "display": "inline",
    "initialValue": 2021-11-09T23:00:00.000Z,
    "labelExtractor": [Function],
    "locale": undefined,
    "mode": "date",
    "trim": false,
    "utc": false,
  },
  "defaultStyles": Object {
    "disabledText": Object {
      "color": "#cdcdcd",
    },
    "disabledView": Object {},
    "text": Object {
      "color": "#007aff",
      "fontSize": 20,
    },
    "view": Object {
      "alignItems": "center",
      "flexDirection": "row",
      "flexGrow": 1,
      "height": 56,
      "justifyContent": "center",
    },
  },
  "inputButtonDefaultProps": Object {
    "placeholderTextColor": "#c7c7cc",
  },
  "inputButtonDefaultStyles": Object {
    "disabledText": Object {
      "color": "#cdcdcd",
    },
    "disabledView": Object {},
    "text": Object {
      "color": "black",
      "fontSize": 14,
    },
    "view": Object {
      "alignItems": "center",
      "justifyContent": "flex-start",
    },
  },
  "modalDialogDefaultProps": Object {
    "animationType": "slide",
    "backgroundColor": "white",
    "cancelTitle": "Cancel",
    "confirmTitle": "Confirm",
    "transparent": true,
  },
  "modalDialogDefaultStyles": Object {
    "body": Object {
      "flexDirection": "column",
      "flexShrink": 1,
    },
    "cancel": Object {
      "borderRadius": 12,
      "fontWeight": "600",
      "marginTop": 24,
    },
    "confirm": Object {
      "borderBottomLeftRadius": 12,
      "borderBottomRightRadius": 12,
      "fontWeight": "400",
    },
    "container": Object {
      "alignItems": "stretch",
      "flexDirection": "column",
      "marginBottom": 8,
      "marginHorizontal": 8,
      "maxHeight": 1326,
    },
    "footer": Object {
      "borderBottomLeftRadius": 12,
      "borderBottomRightRadius": 12,
    },
    "header": Object {
      "alignItems": "center",
      "borderTopLeftRadius": 12,
      "borderTopRightRadius": 12,
      "padding": 16,
    },
    "message": Object {
      "color": "#888",
      "fontSize": 13,
      "fontWeight": "400",
      "paddingBottom": 12,
      "textAlign": "center",
    },
    "modal": Object {
      "flex": 1,
      "justifyContent": "flex-end",
    },
    "title": Object {
      "color": "#888",
      "fontSize": 14,
      "fontWeight": "500",
      "paddingBottom": 12,
      "textAlign": "center",
    },
  },
  "monthPickerDefaultProps": Object {
    "InputButtonComponent": [Function],
    "initialValue": 2021-11-01T00:00:00.000Z,
    "labelExtractor": [Function],
    "locale": undefined,
    "trim": false,
    "utc": false,
  },
  "pressableDefaultProps": Object {
    "TouchableComponent": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
  },
  "pressableDefaultStyles": Object {
    "disabledText": Object {
      "color": "#cdcdcd",
    },
    "disabledView": Object {},
    "text": Object {
      "alignSelf": "center",
      "color": "#007aff",
      "fontSize": 18,
    },
    "view": Object {
      "flexDirection": "row",
      "height": 50,
      "justifyContent": "center",
    },
  },
  "selectDefaultProps": Object {
    "InputButtonComponent": [Function],
  },
  "yearPickerDefaultProps": Object {
    "InputButtonComponent": [Function],
    "initialValue": 2021,
    "labelExtractor": [Function],
    "locale": undefined,
    "maxValue": 2041,
    "minValue": 1971,
  },
}
`;
